portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -61,15 +60,26 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    List<DataComment> allComments = new ArrayList<>();
+    int numComment = 0;
+    // parse max number of comments from user input
+    if (request.getParameter("max-comments") != null) {
+        String maxCommentsString = request.getParameter("max-comments");
+        maxComments = Integer.parseInt(maxCommentsString);
+    }
+
+    // limit number of comments shown
+    List<DataComment> comments = new ArrayList<>();
     for (Entity entity : results.asIterable()) {
       String text = (String) entity.getProperty("text");
-      DataComment newComment = new DataComment(text);
-      allComments.add(newComment);
+      DataComment currComment = new DataComment(text);
+      if (numComment < maxComments) {
+        comments.add(currComment);
+        numComment++; 
```

Comment:
There should be a break here, otherwise you'll retrieve the rest of the comments from Datastore, which will be immediately thrown away. This is also a good candidate for extracting to a helper function:

`private List<DataComment> comments = retrieveComments(datastore, maxComments);`